Core/CwCheat.cpp

```diff
@@ -158,608 +313,709 @@ void CWCheatEngine::CreateCheatFile() {
+	CheatOp op;
+	uint32_t addr;
+	int sz;
+	uint32_t val;
+
+	union {
+		struct {
+			uint32_t count;
+			uint32_t step;
+			uint32_t add;
+		} multiWrite;
+		struct {
+			uint32_t destAddr;
+		} copyBytesFrom;
+		struct {
+			uint32_t skip;
+		} ifTypes;
+		struct {
+			uint32_t skip;
+			uint32_t compareAddr;
+		} ifAddrTypes;
+		struct {
+			int offset;
+			int baseOffset;
+			int count;
+			int type;
+		} pointerCommands;
+	};
+};
+
+CheatOperation CWCheatEngine::InterpretNextCwCheat(const CheatCode &cheat, size_t &i) {
+	const CheatLine &line1 = cheat.lines[i++];
+	const uint32_t &arg = line1.part2;
+
+	// Filled as needed.
+	u32 addr;
+
+	int type = line1.part1 >> 28;
+	switch (type) {
+	case 0x0: // Write 8-bit data (up to 4 bytes.)
+		addr = GetAddress(line1.part1 & 0x0FFFFFFF);
+		if (arg & 0xFFFF0000)
+			return { CheatOp::Write, addr, 4, arg };
+		else if (arg & 0x0000FF00)
+			return { CheatOp::Write, addr, 2, arg };
+		else
+			return { CheatOp::Write, addr, 1, arg };
+
+	case 0x1: // Write 16-bit data.
+		addr = GetAddress(line1.part1 & 0x0FFFFFFF);
+		return { CheatOp::Write, addr, 2, arg };
+
+	case 0x2: // Write 32-bit data.
+		addr = GetAddress(line1.part1 & 0x0FFFFFFF);
+		return { CheatOp::Write, addr, 4, arg };
+
+	case 0x3: // Increment/decrement data.
+		addr = GetAddress(arg & 0x0FFFFFFF);
+		switch ((line1.part1 >> 20) & 0xF) {
+		case 1:
+			return { CheatOp::Add, addr, 1, line1.part1 & 0xFF };
+		case 2:
+			return { CheatOp::Subtract, addr, 1, line1.part1 & 0xFF };
+		case 3:
+			return { CheatOp::Add, addr, 2, line1.part1 & 0xFFFF };
+		case 4:
+			return { CheatOp::Subtract, addr, 2, line1.part1 & 0xFFFF };
+		case 5:
+			if (i < cheat.lines.size())
+				return { CheatOp::Add, addr, 4, cheat.lines[i++].part1 };
+			return { CheatOp::Invalid };
+		case 6:
+			if (i < cheat.lines.size())
+				return { CheatOp::Subtract, addr, 4, cheat.lines[i++].part1 };
+			return { CheatOp::Invalid };
+
+		default:
+			return { CheatOp::Invalid };
+		}
+		break;
+
+	case 0x4: // 32-bit multi-write patch data.
+		addr = GetAddress(line1.part1 & 0x0FFFFFFF);
+		if (i < cheat.lines.size()) {
+			const CheatLine &line2 = cheat.lines[i++];
+
+			CheatOperation op = { CheatOp::MultiWrite, addr, 4, line2.part1 };
+			op.multiWrite.count = arg >> 16;
+			op.multiWrite.step = (arg & 0xFFFF) * 4;
+			op.multiWrite.add = line2.part2;
+			return op;
+		}
+		return { CheatOp::Invalid };
+
+	case 0x5: // Memcpy command.
+		addr = GetAddress(line1.part1 & 0x0FFFFFFF);
+		if (i < cheat.lines.size()) {
+			const CheatLine &line2 = cheat.lines[i++];
+
+			CheatOperation op = { CheatOp::CopyBytesFrom, addr, 0, arg };
+			op.copyBytesFrom.destAddr = GetAddress(line2.part1 & 0x0FFFFFFF);
+			return op;
+		}
+		return { CheatOp::Invalid };
+
+	case 0x6: // Pointer commands.
+		addr = GetAddress(line1.part1 & 0x0FFFFFFF);
+		if (i < cheat.lines.size()) {
+			const CheatLine &line2 = cheat.lines[i++];
+			const int count = line1.part1 & 0xFFFF;
```

Comment:
This is actually taking address part of the pointer code and because of that hitting not enough lines below breaking those annoying 0x6 code types:].